src/cpp/server/load_reporter/load_reporter_async_service_impl.cc

```diff
@@ -0,0 +1,357 @@
+                               gpr_time_from_seconds(1, GPR_TIMESPAN)));
+  std::shared_ptr<ReportLoadHandler> handler =
+      std::make_shared<ReportLoadHandler>(service->cq_.get(), service,
+                                          service->load_reporter_.get());
+  handler->Start();
+  // Orphan the handler.
+  handler.reset();
+  void* tag;
+  bool ok;
+  while (true) {
+    if (!service->cq_->Next(&tag, &ok)) {
+      // The completion queue is shutting down.
+      GPR_ASSERT(service->shutdown_);
+      break;
+    }
+    auto next_step = static_cast<std::function<void(bool)>*>(tag);
+    (*next_step)(ok);
+  }
+}
+
+void LoadReporterAsyncServiceImpl::StartThread() { thread_->Start(); }
+
+LoadReporterAsyncServiceImpl::ReportLoadHandler::ReportLoadHandler(
+    ServerCompletionQueue* cq, LoadReporterAsyncServiceImpl* service,
+    LoadReporter* load_reporter)
+    : cq_(cq),
+      service_(service),
+      load_reporter_(load_reporter),
+      stream_(&ctx_),
+      call_status_(WAITING_FOR_DELIVERY) {}
+
+void LoadReporterAsyncServiceImpl::ReportLoadHandler::Start() {
+  std::unique_lock<std::mutex> lock(service_->cq_shutdown_mu_);
+  if (service_->shutdown_) return;
+  // AsyncNotifyWhenDone() needs to be called before the call starts, but the
+  // tag may not pop out if the call never starts. Please refer to
+  // https://github.com/grpc/grpc/issues/10136.
+  on_done_notified_ = std::function<void(bool)>(
+      std::bind(&ReportLoadHandler::OnDoneNotified, shared_from_this(),
+                &on_done_notified_, std::placeholders::_1));
+  next_inbound_ = std::function<void(bool)>(
+      std::bind(&ReportLoadHandler::OnRequestDelivered, shared_from_this(),
+                &next_inbound_, std::placeholders::_1));
+  ctx_.AsyncNotifyWhenDone(&on_done_notified_);
+  service_->RequestReportLoad(&ctx_, &stream_, cq_, cq_, &next_inbound_);
+}
+
+void LoadReporterAsyncServiceImpl::ReportLoadHandler::OnRequestDelivered(
+    std::function<void(bool)>* func, bool ok) {
+  if (ok) call_status_ = DELIVERED;
+  if (!ok || shutdown_) {
+    // The value of ok being false means that the server is shutting down.
+    Shutdown(func, "OnRequestDelivered");
+    return;
+  }
+  // Spawn a new handler instance to serve the next new client. Every handler
+  // instance will deallocate itself when it's done.
+  std::shared_ptr<ReportLoadHandler> handler =
+      std::make_shared<ReportLoadHandler>(cq_, service_, load_reporter_);
+  handler->Start();
+  // Orphan the handler.
+  handler.reset();
+  *func = std::function<void(bool)>(std::bind(&ReportLoadHandler::OnReadDone,
+                                              shared_from_this(), func,
+                                              std::placeholders::_1));
+  {
+    std::unique_lock<std::mutex> lock(service_->cq_shutdown_mu_);
+    if (service_->shutdown_) {
+      lock.release()->unlock();
+      Shutdown(func, "OnRequestDelivered");
+      return;
+    }
+    stream_.Read(&request_, func);
```

Comment:
nit: It is better to have comment saying we only expect one request. And the unexpected request will trigger OnReadDone.